[1.12, 1.13 regression] `typing.overload` and `ParamSpec`

<!--
If you're not sure whether what you're experiencing is a mypy bug, please see the "Question and Help" form instead.
Please consider:

- checking our common issues page: https://mypy.readthedocs.io/en/stable/common_issues.html
- searching our issue tracker: https://github.com/python/mypy/issues to see if it's already been reported
- asking on gitter chat: https://gitter.im/python/typing
-->

**Bug Report**

There are two overloaded versions of function: the first version for `AsyncContextManager` parameter, the second version for `Callable` with ParamSpec. 
Static type checker must choose the first version, but it chooses the second one.    

BUT: after removing `ParamSpec` from function signature, mypy ends up working well (see the comment in example).

Note: pyright (v. 1.1.386) works just fine on it.
Possibly related to #17960 

**To Reproduce**

```python
from __future__ import annotations

from types import TracebackType
from typing import Any, AsyncContextManager, Callable, TypeVar, overload

from typing_extensions import ParamSpec, Self

T = TypeVar("T")
P = ParamSpec("P")


class Animal:
    async def __aenter__(self) -> Self:
        return self

    async def __aexit__(
        self,
        typ: type[BaseException] | None,
        exc: BaseException | None,
        tb: TracebackType | None,
    ) -> None:
        return None

    def __call__(self) -> str:
        return "name"


@overload
def name(animal: AsyncContextManager[T]) -> T: ...


@overload
def name(
    animal: Callable[P, T],
    *_args: P.args,
    **_kwargs: P.kwargs,
) -> T: ...


# BUT!
# if replace the second overload with
# @overload
# def name(animal: Callable[[], T]) -> T: ...
# or even
# def name(animal: Callable[[int], T]) -> T: ...
# everything works fine


def name(
    animal: Any,
    *_args: Any,
    **_kwargs: Any,
) -> Any:
    return animal


dog = Animal()
dog2: Animal = name(dog)
```

[Live playground link](https://mypy-play.net/?mypy=latest&python=3.12&gist=e7cd0c36dbb62796cad89e12da0d540f)

**Expected Behavior**
mypy has not reported any issue with such overloaded functions, and should still not report as it was for mypy prior to 1.12.0.

**Actual Behavior**
```bash
main.py:58: error: Incompatible types in assignment (expression has type "str", variable has type "Animal")  [assignment]
Found 1 error in 1 file (checked 1 source file)
```


**Your Environment**

- Mypy version used: 1.12.0, 1.12.1, 1.13.0
- Mypy command-line flags: N/A
- Mypy configuration options from `mypy.ini` (and other config files): N/A
- Python version used: 3.8, 3.9, 3.10, 3.11, 3.12

--- a/mypy/checkexpr.py
+++ b/mypy/checkexpr.py
@@ -2781,7 +2781,7 @@ class ExpressionChecker(ExpressionVisitor[Type]):
     ) -> list[CallableType]:
         """Returns all overload call targets that having matching argument counts.
 
-        If the given args contains a star-arg (*arg or **kwarg argument, including
+        If the given args contains a star-arg (*arg or **kwarg argument, except for
         ParamSpec), this method will ensure all star-arg overloads appear at the start
         of the list, instead of their usual location.
 
@@ -2816,7 +2816,9 @@ class ExpressionChecker(ExpressionVisitor[Type]):
                     # ParamSpec can be expanded in a lot of different ways. We may try
                     # to expand it here instead, but picking an impossible overload
                     # is safe: it will be filtered out later.
-                    star_matches.append(typ)
+                    # Unlike other var-args signatures, ParamSpec produces essentially
+                    # a fixed signature, so there's no need to push them to the top.
+                    matches.append(typ)
                 elif self.check_argument_count(
                     typ, arg_types, arg_kinds, arg_names, formal_to_actual, None
                 ):

--- a/test-data/unit/check-parameter-specification.test
+++ b/test-data/unit/check-parameter-specification.test
@@ -2303,3 +2303,38 @@ reveal_type(capture(fn))  # N: Revealed type is "Union[builtins.str, builtins.in
 reveal_type(capture(err))  # N: Revealed type is "builtins.int"
 
 [builtins fixtures/paramspec.pyi]
+
+[case testRunParamSpecOverlappingOverloadsOrder]
+from typing import Any, Callable, overload
+from typing_extensions import ParamSpec
+
+P = ParamSpec("P")
+
+class Base:
+    pass
+class Child(Base):
+    def __call__(self) -> str: ...
+class NotChild:
+    def __call__(self) -> str: ...
+
+@overload
+def handle(func: Base) -> int: ...
+@overload
+def handle(func: Callable[P, str], *args: P.args, **kwargs: P.kwargs) -> str: ...
+def handle(func: Any, *args: Any, **kwargs: Any) -> Any:
+    return func(*args, **kwargs)
+
+@overload
+def handle_reversed(func: Callable[P, str], *args: P.args, **kwargs: P.kwargs) -> str: ...
+@overload
+def handle_reversed(func: Base) -> int: ...
+def handle_reversed(func: Any, *args: Any, **kwargs: Any) -> Any:
+    return func(*args, **kwargs)
+
+reveal_type(handle(Child()))  # N: Revealed type is "builtins.int"
+reveal_type(handle(NotChild()))  # N: Revealed type is "builtins.str"
+
+reveal_type(handle_reversed(Child()))  # N: Revealed type is "builtins.str"
+reveal_type(handle_reversed(NotChild()))  # N: Revealed type is "builtins.str"
+
+[builtins fixtures/paramspec.pyi]

